### PR TITLE
Ajaxify appliances

### PR DESCRIPTION
**Pull request by cbulzoni**
I added Ajax to create, update and destroy for the appliances model. I also made some formatting tweaks but I am more interested in feedback around Ajax.

### CODE REVIEW

File: app/views/appliances/_form.html.erb [hunk 1]
```diff
@@ -1,57 +1,59 @@
...
+    </div>
 
-      <ul>
-        <% appliance.errors.each do |error| %>
-          <li><%= error.full_message %></li>
-        <% end %>
-      </ul>
+    <div class="field">
+      <%= form.label :appliance_type %>
+      <%= form.select :appliance_type, @possible_appliance_types.map {|value| [value,value]}, :prompt => 'Select One' %>
```
Comment (jelaniwoods) [inline app/views/appliances/_form.html.erb:22]: Creating an instance variable for the select option feels a little inconvenient to do each time you want to use this form.

You could add an Inclusion validation or even a Enum to make this more dynamic.

If you update your model

```rb
class Appliance < ApplicationRecord
  belongs_to :house
  belongs_to :contact, required: false
  has_one :owner, through: :house, source: :owner

  validates :appliance_type, inclusion: ["Air Conditioner","Water Heater","Sump Pump","Furnace","Range","Oven","Stove","Microwave","Dishwasher","Washing Machine","Dryer","Other"]
end
```

Then in this form you can do something like

```suggestion
      <%= form.select :appliance_type,  Appliance.validators.last.options[:in], :prompt => 'Select One' %>
```

Note: `.last` works here because it's the last validation defined in your model. If you add more validations you should consider adding them _above_ this one.

File: app/views/appliances/_form.html.erb [hunk 2]
```diff
@@ -1,57 +1,59 @@
...
+    <div class="field">
+      <%= form.label :contact_id %>
+      <%= form.select :contact_id, current_user.contacts.collect {|c| [c.contact_name, c.id] }, include_blank: true %>
+    </div>
+
+    <div class="actions">
+      <%= form.submit %>
+    </div>
+  <% end %>
+</div>
```
Comment (jelaniwoods) [inline app/views/appliances/_form.html.erb:59]: It might be nice to add a remote `link_to` to replace the form with the new appliance plus button, in case the user changes their mind about adding an appliance. Currently a full page refresh is required.

File: app/views/appliances/create.js.erb [hunk 3]
```diff
@@ -0,0 +1 @@
+$("#new_appliance_form").replaceWith("<%= j(render @appliance) %>");
```
Comment (jelaniwoods) [inline app/views/appliances/create.js.erb]: This works, but maybe isn't the best experience since after you create one Appliance, the form disappears from the page and you have to reload to get the form back.

A solution to this is a little tricky since the form is in the same line as the the first two appliances 🤔 

You could try ordering the appliances on the index so the new appliance appears at the end of the list. You couldn't use `replaceWith` but could maybe use [append()](https://api.jquery.com/append/) as long as you identify the last row of Appliances on the page.

This also might not be the user experience you want though, since the new appliance will be at the bottom of the page and then you'd have to scroll to see it 😓 

Perhaps the easiest solution is to create a View partial that is responsible for rendering _all_ appliances as well as the new button/form.

Basically if this was a partial

https://github.com/cbulzoni-appdev/house-manager-v2-rails/blob/68bf8303b48e6c474101d9382cf3f9ac1309624f/app/views/appliances/index.html.erb#L7-L33

then you could put an id on it and re-render this whole partial with the new list of appliances.

This is pretty nitpicky— what do you think?

File: app/views/appliances/destroy.js.erb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+$("#<%= dom_id(@appliance) %>").fadeOut(function() {
+  $(this).remove();
+});
```
Comment (jelaniwoods) [inline app/views/appliances/destroy.js.erb:3]: I think this isn't removing everything it needs to. The id of the appliance is on the card, but it's _not_ on the div column— meaning if I delete an appliance in the middle of a row, the empty space remains.

remove-appliance

So you could put an id on the div column instead or use a different selector, but this can still run into issues if appliances in two different rows are deleted: 

Screen Shot 2022-06-03 at 11 26 06 AM

So I think actually the better solution is to use one div row for **every** appliance, including the form. Bootstrap column classes are smart enough so that if more than 3 `col-md-4`s appear in one row, the fourth column will appear on the next line. More on that in the final review comment.

**Review by jelaniwoods — COMMENTED** [review]
@cbulzoni the AJAX definitely works— I think the user experience of the AJAX could be improved some. Right now it's difficult to create or delete appliances from the index page smoothly, because there are two different div rows where an appliance can appear.

I think to make things easier to "ajax-ify" you can restructure the HTML for the appliances#index page to use one div row for **every** appliance, including the form—

something like:

```erb
    <div class="row mb-4">
      <div class="col-md-4">
        <div class="card" style="width: 18em;" id="new_appliance_button">
          <%= link_to new_appliance_path, class: "card-image-top", remote: true do %>
            <h5>Add New Appliance</h5>
            <img src="https://media.istockphoto.com/vectors/black-plus-sign-positive-symbol-vector-id688550958?k=20&m=688550958&s=612x612&w=0&h=wvzUqT3u3feYygOXg3GB9pYBbqIsyu_xpvfTX-6HOd0=" class="card-img-bottom">
          <% end %>
        </div>
      </div>
      <% @appliances.each do |appliance| %>
        <div class="col-md-4">
          <%= render "appliances/appliance", appliance: appliance %>
        </div>
      <% end %>
    </div>
```
_note: I might also remove the `style="width: ..."` attributes from your bootstrap card components on the index and appliance partial. That's why they overflow when the screen gets small_

With this new structure it's a lot easier to add a partial directly next to the form or at the end of the list using different css selectors of JQuery functions.

What do you think?
